src/plugins/workspace/public/plugin.ts

```diff
@@ -2,14 +2,78 @@
  * Copyright OpenSearch Contributors
  * SPDX-License-Identifier: Apache-2.0
  */
-import { Plugin } from '../../../core/public';
 
-export class WorkspacePlugin implements Plugin<{}, {}, {}> {
-  public async setup() {
-    return {};
+import { i18n } from '@osd/i18n';
+import { combineLatest } from 'rxjs';
+import { map } from 'rxjs/operators';
+import {
+  ChromeNavLink,
+  CoreSetup,
+  CoreStart,
+  Plugin,
+  WorkspaceObject,
+  DEFAULT_APP_CATEGORIES,
+} from '../../../core/public';
+import { SavedObjectsManagementPluginSetup } from '../../saved_objects_management/public';
+import { featureMatchesConfig } from './utils';
+
+interface WorkspacesPluginSetupDeps {
+  savedObjectsManagement?: SavedObjectsManagementPluginSetup;
+}
+
+export class WorkspacePlugin implements Plugin<{}, {}, WorkspacesPluginSetupDeps> {
+  private filterByWorkspace(workspace: WorkspaceObject | null, allNavLinks: ChromeNavLink[]) {
+    if (!workspace) return allNavLinks;
+    const features = workspace.features ?? ['*'];
+    return allNavLinks.filter(featureMatchesConfig(features));
   }
 
-  public start() {
+  private filterNavLinks(core: CoreStart) {
+    const navLinksService = core.chrome.navLinks;
+    const chromeNavLinks$ = navLinksService.getNavLinks$();
+    const currentWorkspace$ = core.workspaces.currentWorkspace$;
+    combineLatest([
+      chromeNavLinks$.pipe(map(this.changeCategoryNameByWorkspaceFeatureFlag)),
+      currentWorkspace$,
+    ]).subscribe(([chromeNavLinks, currentWorkspace]) => {
+      const filteredNavLinks = new Map<string, ChromeNavLink>();
+      chromeNavLinks = this.filterByWorkspace(currentWorkspace, chromeNavLinks);
+      chromeNavLinks.forEach((chromeNavLink) => {
+        filteredNavLinks.set(chromeNavLink.id, chromeNavLink);
+      });
+      navLinksService.setFilteredNavLinks(filteredNavLinks);
+    });
+  }
+
+  /**
+   * The category "Opensearch Dashboards" needs to be renamed as "Library"
+   * when workspace feature flag is on, we need to do it here and generate
+   * a new item without polluting the original ChromeNavLink.
+   */
+  private changeCategoryNameByWorkspaceFeatureFlag(chromeLinks: ChromeNavLink[]): ChromeNavLink[] {
+    return chromeLinks.map((item) => {
+      if (item.category?.id === DEFAULT_APP_CATEGORIES.opensearchDashboards.id) {
+        return {
+          ...item,
+          category: {
+            ...item.category,
+            label: i18n.translate('core.ui.libraryNavList.label', {
+              defaultMessage: 'Library',
+            }),
+          },
+        };
+      }
+      return item;
+    });
+  }
+
+  public async setup(core: CoreSetup) {
+    core.workspaces.workspaceEnabled$.next(true);
+  }
+  public start(core: CoreStart) {
+    if (core) {
+      this.filterNavLinks(core);
```

Comment:
Prefer add some unit tests for setting filtered nav links.

Reply:
Ok, I will try to test this `start` function.

Reply:
Ok, I will try to test this `start` function.